doc/designs/quic-design/dgram-api.md

```diff
@@ -0,0 +1,322 @@
+Datagram BIO API revisions for sendmmsg/recvmmsg
+================================================
+
+We need to evolve the API surface of BIO which is relevant to BIO_dgram (and the
+eventual BIO_dgram_mem) to support APIs which allow multiple datagrams to be
+sent or received simultaneously, such as sendmmsg(2)/recvmmsg(2).
```

Comment:
Supporting iovec creates an issue when trying to implement it on OSes that don't support it. At that point one's only option is to copy into a staging buffer (since this is UDP, and unlike TCP multiple sequential send() calls don't  have the same semantics as one call).

Such copying would give these functions a very surprising performance profile to applications that use them, where they are fast on some systems and slow on another. Rather than do this, we expose an API which we can support performantly on all platforms.

The primary use case for this API is for our QUIC implementation. The basic view is that lack of iovec here is not actually going to be a performance issue because the only things sent and received using this API by QUIC are going to be encrypted QUIC packets. This means you are copying, potentially from multiple places, into a staging buffer anyway (and encrypting in the process). The only use for iovec here is maybe to concatenate the QUIC packet header and the encrypted packet body, but it is easier just to have our code which generates encrypted QUIC packets leave room for the QUIC packet header in the staging buffer containing the encrypted data and fill it in there. This avoids the need for iovec support with the OS support issues that would create, while remaining maximally performant.